OpenRA.Mods.Common/Traits/World/ClusterManager.cs

```diff
@@ -0,0 +1,314 @@
+
+			while (toProcess.Count != 0)
+			{
+				var start = toProcess.Dequeue();
+				if (!toCheck.Contains(start))
+					continue;
+
+				var clusterQueue = new Queue<CPos>();
+				clusterQueue.Enqueue(start);
+
+				while (clusterQueue.Count != 0)
+				{
+					var n = clusterQueue.Dequeue();
+					if (!toCheck.Contains(n))
+						continue;
+
+					toCheck.Remove(n);
+					UpdateCell(n, selectedCluster);
+
+					CVec.Directions.Select(d => d + n)
+						.Where(d => toCheck.Contains(d))
+						.Do(d => clusterQueue.Enqueue(d));
+				}
+
+				// Don't create a new cluster if all cells are checked
+				if (toCheck.Count != 0)
+				{
+					if (selectedCluster == cluster)
+					{
+						cluster.Cells.RemoveAll(x => toCheck.Contains(x));
+					}
+
+					selectedCluster = new ClusterContents();
+					clusters.Add(selectedCluster);
+				}
+			}
+		}
+
+		protected virtual ushort BuildDomain()
+		{
+			ClusterContents cluster = null;
+
+			var visited = new CellLayer<bool>(map);
+
+			var toProcess = new Queue<CPos>();
+			toProcess.Enqueue(MPos.Zero.ToCPos(map));
+
+			// Flood-fill over each cluster.
+			while (toProcess.Count != 0)
+			{
+				var start = toProcess.Dequeue();
+
+				// Technically redundant with the check in the inner loop, but prevents
+				// ballooning the cluster amount.
+				if (visited[start])
+					continue;
+
+				var clusterQueue = new Queue<CPos>();
+				clusterQueue.Enqueue(start);
+
+				var currentTrue = ClusterCondition(start);
+				cluster = new ClusterContents();
+				clusters.Add(cluster);
+
+				// Add all contiguous cells to our cluster, and make a note of
+				// any non-contiguous cells for future clusters.
+				while (clusterQueue.Count != 0)
+				{
+					var n = clusterQueue.Dequeue();
+					if (visited[n])
+						continue;
+
+					// add cell to queue if cell does not belong in queue
+					if (currentTrue != ClusterCondition(n))
+					{
+						toProcess.Enqueue(n);
+						continue;
+					}
+
+					visited[n] = true;
+					if (currentTrue)
+					{
+						clusterCellLayer[n] = cluster;
+						cluster.Cells.Add(n);
+					}
+					else
+						clusterCellLayer[n] = emptyCluster;
+
+					// Don't crawl off the map, or add already-visited cells.
+					CVec.Directions.Select(d => d + n)
+						.Where(d => visited.Contains(d) && !visited[d])
+						.Do(d => clusterQueue.Enqueue(d));
+				}
+			}
+
+			return cluster.ClusterId;
+		}
+	}
+
+	class ClusterContents
+	{
+		public readonly bool IsEmpty;
+
+		// For debugging only
+		static ushort domainIndex = 1;
+		public readonly ushort ClusterId;
+		public List<CPos> Cells = new List<CPos>();
+
+		public readonly List<ClusterTransientConnection> TransientConnections = new List<ClusterTransientConnection>();
+
+		public ClusterContents()
+		{
+			ClusterId = domainIndex;
+			domainIndex += 1;
+		}
+
+		public ClusterContents(bool empty)
+		{
+			ClusterId = 0;
+			IsEmpty = true;
+		}
+	}
+
+	// For TS bridges and tunnels
+	class ClusterTransientConnection
```

Comment:
I still haven't tested if they are updated dynamically, I need to create a good testcase. 
Another note, since these will likely be only used for TS bridges and tunnels should I extract this code to DomainIndex?